rfcs/012-add-session-ssh-support/012-add-session-ssh-support.md

```diff
@@ -0,0 +1,438 @@
+configuration and getting access to ssh services must be clear for both
+contexts and map to the design of the different platforms
+
+## Key Assumptions
+
+The following key assumptions apply:
+- we are not targeting the most basic users; we assume they understand the
+  need for some security and are willing to put the time/energy required into a
+  reasonably sensible ssh configuration
+- password authentication to ssh sessions exposed to the Internet is not
+  permitted
+
+## Possible solutions
+
+As this work has some kind of history, three solutions have been discussed in
+various conversations - those are the first three options below; a fourth
+approach arose when presenting these ideas to the team and that is the last
+item in the list below:
+- an approach in which an ssh port is exposed on each session directly to the
+  Internet;
+- an approach in which a dedicated jumphost is used and the standard ssh
+  proxying mechanisms are used to access the session via the jumphost;
+- an approach in which a dedicated proxy is provided as part of the Renku
+  platform which provides ssh access without needing to use the ssh client
+  proxy capabilities.
+- an approach which uses OAuth/OIDC to enable login without using either keys
+  or long-term passwords
+
+Each of these approaches is discussed in more detail below.
+
+Other solutions could be envisaged: it could be possible, for example, to
+provide a tunnel over HTTP/S into the session and access it via ssh locally or
+perhaps some wireguard solution might work but these approaches would probably
+encounter issues with different OS versions, would not be easy to configure and
+ultimately would likely have a detrimental impact on user experience and hence
+are not considered further.
+
+### Exposing ssh port from each session directly
+
+In this approach, each user session pod exposes at least 2 ports: an HTTP port
+for the Jupyter lab server and an ssh port for the ssh session. These ports
+have corresponding services running inside the pod.
+
+Kubernetes services are then linked to these sessions with the ports on the
+services mapping to the ports exposed by the pod. An ingress is then mapped to
+the ssh services with a dedicated ingress for each ssh session.
+
+Each ingress should have a unique name and also a unique IP address as the ssh
+protocol operates on an IP address basis (i.e., it has no specific server
+identification as noted above). This solution can suffer from DNS propagation
+delays, depending on the DNS provider and DNS configuration. Further, exposing
+ssh servers running inside user sessions directly to the Internet brings some
+security risks and it would need to be clear that users cannot easily change
+the ssh daemon configuration within their sessions to make their session very
+exposed.
+
+A couple of further points relating to this approach:
+- ingresses on cloud providers can typically be expensive; having one per
```

Comment:
do you mean here specifically a load-balancer ingress? When I read "ingress" I think the k8s ingress, which may or may not be a load-balancer.

Reply:
Good point - I will clarify.